### PR TITLE
fix(subscription): codex official Keychain fallback + OAuth path routing

### LINKED ISSUE #2157: 今天打开 cc switch后，codex得官方渠道上一直有一个 会话已过期，无法消除
### Self Checks / 自检

- [x] I have read the [FAQ](https://github.com/farion1231/cc-switch#faq) section in README.
我已阅读 README 中的[常见问题](https://github.com/farion1231/cc-switch#常见问题)。

- [x] I have searched for [existing issues](https://github.com/farion1231/cc-switch/issues), including closed ones.
我已搜索过[已有的 Issue](https://github.com/farion1231/cc-switch/issues)，包括已关闭的。

### CC Switch Version / 版本号

3.13.0

### Operating System / 操作系统

macOS

### Related App / 涉及应用

Codex

### Steps to Reproduce / 重现步骤

Image

### Expected Behavior / 期望行为

希望可以移除这个提示，我重试了很多次重新登录codex，也卸载了app，重新安装了很多次，一直都在

### Actual Behavior / 实际行为

_No response_

### Additional Context / 补充信息

_No response_

### LINKED ISSUE #3479: Codex Official quota check should not let stale macOS Keychain Codex Auth shadow ~/.codex/auth.json
## Summary

On macOS, the Codex Official quota check currently appears to prefer the macOS Keychain entry `Codex Auth` over `~/.codex/auth.json`. If an old/stale Keychain entry exists, CC Switch shows `会话已过期` / `Session expired`, even when the current Codex CLI auth file is valid and can query the official quota endpoint successfully.

This looks like a credential selection / fallback issue in the Codex Official subscription quota path.

## Environment

- CC Switch: v3.16.0
- macOS
- Codex CLI: `codex-cli 0.135.0`
- Provider: Codex tab -> `OpenAI Official`

## Observed behavior

The `OpenAI Official` card displayed `会话已过期`.

Local diagnostics, with secrets redacted:

```text
Keychain service "Codex Auth":
  last_refresh: 2026-04-14T03:27:02Z
  wham/usage: HTTP 401 token_expired

~/.codex/auth.json:
  last_refresh: 2026-05-31T23:38:59Z
  wham/usage: HTTP 200
  has_rate_limit: true
```

After deleting stale Keychain `Codex Auth` records, CC Switch immediately recovered and displayed the Codex quota correctly:

```text
5h window: 81%
7d window: 19%
```

No token, email, or account id was changed in CC Switch. The only change was removing stale `Codex Auth` entries from the macOS Keychain so the valid `~/.codex/auth.json` could be used.

## Expected behavior

If `~/.codex/auth.json` contains a valid current Codex CLI login, Codex Official quota should use it or at least fall back to it when the Keychain token is stale/invalid.

A stale Keychain record should not shadow a newer valid `~/.codex/auth.json`.

## Why this seems wrong

Current OpenAI Codex source appears to define CLI auth file storage as the default in current code:

- `AuthCredentialsStoreMode::File` is the default
- `File` means credentials are stored in `CODEX_HOME/auth.json`
- `CODEX_HOME` defaults to `~/.codex`

Historically, Codex had commits around 2026-02 that experimented with `keyring` / `auto` defaults, but current source and this local Codex CLI installation are using `~/.codex/auth.json` as the effective current credential store.

So a legacy/stale `Codex Auth` Keychain item can remain on disk and break CC Switch quota display even though Codex itself is logged in and working.

## Current CC Switch code path I inspected

In `src-tauri/src/services/subscription.rs`, `read_codex_credentials()` tries:

1. macOS Keychain service `Codex Auth`
2. `~/.codex/auth.json`

The problem is that if Keychain returns a stale auth object, CC Switch uses it directly. Then `query_codex_quota()` calls:

```text
https://chatgpt.com/backend-api/wham/usage
```

If the stale token returns `401` or `403`, CC Switch marks the credential as `Expired`, and the UI shows `会话已过期`. It does not retry `~/.codex/auth.json`.

## Suggested fix

Any of these would fix the issue:

1. Prefer `~/.codex/auth.json` for Codex Official quota checks, because current Codex defaults to file storage.
2. Read both Keychain and `~/.codex/auth.json`, then choose the newer valid credential by `last_refresh`.
3. If Keychain credential returns `401` / `403`, retry with `~/.codex/auth.json` before returning `CredentialStatus::Expired`.
4. Ignore old/incomplete Keychain entries that do not match the current Codex auth schema, such as missing `auth_mode = "chatgpt"`.

## Related symptom

This may explain some reports like #2157, but the specific failure mode here is not missing `codex_oauth` provider metadata. It is a stale macOS Keychain `Codex Auth` entry shadowing a valid `~/.codex/auth.json` for the Codex Official provider.

### PR DESCRIPTION
**Pull request by loocor**
## Summary

Three fixes for the Codex Official provider's subscription quota display:

### 1. Keychain stale-entry fallback (`subscription.rs`)

`read_codex_credentials()` always preferred macOS Keychain `"Codex Auth"` over
`~/.codex/auth.json`. A legacy entry missing `auth_mode` or with an expired token
(`last_refresh` >8 days) returned `NotFound`/`Expired`, and the valid file was never
consulted. Now only `CredentialStatus::Valid` Keychain results take priority; all
other statuses fall through to the file.

Also propagate the `credential_message` for `NotFound` so the frontend shows the
actual reason instead of the generic "No result returned".

### 2. Preset: route new providers through OAuth path (`codexProviderPresets.ts`)

Add `providerType: "codex_oauth"` and `requiresOAuth: true` to the "OpenAI Official"
codex preset. New providers created from this preset now render `CodexOauthQuotaFooter`
(OAuth path) instead of `SubscriptionQuotaFooter` (CLI credential path).

### 3. Database migration v11→v12 (`schema.rs`)

Patch existing codex providers with `category = "official"` and missing `providerType`
to `"codex_oauth"`. Does not touch `authBinding` — users can bind a ChatGPT account
later via Settings → Auth Center.

## Test plan

- [ ] `cargo build` + `tsc --noEmit` pass
- [ ] Stale Keychain entry (missing `auth_mode` or expired) + valid `~/.codex/auth.json`
      → quota displays correctly (file fallback works)
- [ ] Valid Keychain entry → still preferred over file
- [ ] New "OpenAI Official" provider created → `providerType: "codex_oauth"` in meta,
      renders `CodexOauthQuotaFooter`
- [ ] Existing codex official provider (pre-migration) → v11→v12 migration adds
      `providerType`, renders `CodexOauthQuotaFooter` after upgrade

Fixes #3479
Fixes #2157

### CODE REVIEW

File: src-tauri/src/services/subscription.rs [hunk 1]
```diff
@@ -475,7 +475,12 @@ fn read_codex_credentials() -> CodexCredentials {
     #[cfg(target_os = "macos")]
     {
         if let Some(result) = read_codex_credentials_from_keychain() {
-            return result;
+            // Only use Keychain result if credentials were actually found/parsed.
+            // A stale entry (e.g. missing auth_mode) returns NotFound — fall through
+            // to ~/.codex/auth.json so a valid file is not shadowed (#3479).
+            if result.2 != CredentialStatus::NotFound {
```
[the remark this answers is not in the document]
Comment (loocor) [reply src-tauri/src/services/subscription.rs]: Good catch. Tightened the guard to only return Keychain result when `status == Valid` — both `NotFound` (missing `auth_mode`) and `Expired` (`last_refresh` >8d) now fall through to `~/.codex/auth.json`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by farion1231** [comment]
@codex review

File: src-tauri/src/database/schema.rs [hunk 2]
```diff
@@ -1270,6 +1277,63 @@ impl Database {
...
+        let mut updates = Vec::new();
+        for row in rows {
+            let (id, meta_str) = row.map_err(|e| AppError::Database(e.to_string()))?;
+
+            let Ok(mut meta) = serde_json::from_str::<serde_json::Value>(&meta_str) else {
+                continue;
+            };
+
+            // 只处理 category = "official" 且 providerType 缺失的供应商
+            let is_official = meta.get("category").and_then(|v| v.as_str()) == Some("official");
```
[the remark this answers is not in the document]
Comment (loocor) [reply src-tauri/src/database/schema.rs]: Good catch. `category` is a table column, not inside `meta`. Changed the SQL to `WHERE app_type = 'codex' AND category = 'official'` and removed the `meta.category` check entirely.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by supper-ming** [comment]
I ran into the same Codex official subscription quota failure on macOS with v3.16.3.

Observed locally without exposing any token values:

- macOS Keychain has a `Codex Auth` entry with `tokens.access_token`, but the JSON does not include `auth_mode`.
- `~/.codex/auth.json` contains `auth_mode: "chatgpt"` and can successfully call `https://chatgpt.com/backend-api/wham/usage`.
- Current `main` stops after parsing the Keychain entry as `CredentialStatus::NotFound` (`Codex not using OAuth mode`) and never falls back to the valid file credential, so the UI reports no result for the official subscription test.

This PR's Keychain-only-when-valid fallback matches that failure mode and should fix the case where a stale/incomplete Keychain entry shadows a valid `auth.json`.

A small follow-up suggestion: add regression tests around `read_codex_credentials`/the parser boundary for:

1. unusable Keychain result falls through to file credentials,
2. valid Keychain credentials still take priority,
3. NotFound carries the credential message so the frontend does not show only a generic "no result" error.

**Comment by loocor** [comment]
Closing as obsolete.

The Codex official quota path has moved on since this branch was opened:

- `SCHEMA_VERSION` is now 18, so the v11→v12 `providerType` backfill cannot land as written (v12 already creates the profiles table).
- Codex cards now choose the quota footer by managed-account identity (`native_login` vs `managed_account`), not by forcing `providerType: "codex_oauth"` onto every official provider.
- The OpenAI Official preset already includes `providerType` via #3879, and official subscription display is opt-in.

The Keychain stale-entry fallback for CLI credentials (#3479) never landed on `main`. I am not carrying this branch forward; the capability I still need is maintained separately in CodexHelper.

If anyone still hits a stale macOS `Codex Auth` item shadowing a valid `~/.codex/auth.json`, #3479 remains the right issue. #5386 is a closer starting point than this PR.
